### ISSUE TITLE
PersistentView unusable after upgrade to 1.xer series

### ISSUE THREAD

**Issue by michaelom** [scullxbones/akka-persistence-mongo#79]
Hi scullxbones,

we are in the process of migrating a big application to Akka 2.4.1 with plugin version 1.1.1

Our application uses PersistentView for replaying journal messages. Typically the `auto-update-interval` is set to `1 sec`. Journal sizes vary depending on the consumed event type.
After migrating to Akka 2.4.1 using your migration script we found Mongo performance to be extremely poor, caused by long running queries. CPU load is > 50% after just starting two!!! PersistentViews.

Here's a log snipped:
```
account.events query: { $query: { pid: "Account" }, $orderby: { to: -1 } } planSummary: IXSCAN { pid: 1, from: 1, to: 1 } ntoskip:0 nscanned:95273 nscannedObjects:95273 scanAndOrder:1 keyUpdates:0 writeConflicts:0 numYields:744 nreturned:1 reslen:54 locks:{ Global: { acquireCount: { r: 1490 } }, MMAPV1Journal: { acquireCount: { r: 745 } }, Database: { acquireCount: { r: 745 } }, Collection: { acquireCount: { R: 745 } } } 484ms 
```

And here is the corresponding query for the to the 0.x-series:
```
account.events query:{ $and: [ { pid: "Account" }, { $and: [ { sn: { $gte: 772425 } }, { sn: { $lte: 9223372036854775807 } } ] } ] }
```

It seems the query in the 1.x-series does not restrict the result to a sequence number, meaning there is no upper or lower bound. It fetches **all** events of type Account sorted by property `to` in descending order. 

This is a deal breaker for the upgrade. We need to support around 50 PersistentViews simultaneously with an `auto-update-interval` between 1-10 secs. Moving to the (still experimental) Query Journal API is currently not an option.

Any idea why the behaviour changed in 1.x or how to work around it?

**Comment by michaelom** [comment]
Also, there does not seem to be a index supporting descending order of `to`
```
Indexes for events:
[
    {
        "v" : 1,
        "key" : {
            "_id" : 1
        },
        "name" : "_id_",
        "ns" : "account.events"
    },
    {
        "v" : 1,
        "unique" : true,
        "key" : {
            "pid" : 1,
            "from" : 1,
            "to" : 1
        },
        "name" : "events_index",
        "ns" : "account.events"
    }
```

**Comment by ulrichwinter** [comment]
+1

**Comment by scullxbones** [comment]
So:

* Which driver are you using?
* Has the journal been upgraded (journal auto migrate) already?

That query looks like the `maxSequenceNr` query.  There's a projection (presumably not logged because I see it in the code) that pulls just `to` data, not entire event.  I believe it runs as part of the recovery phase for `PersistentActor`s, and presumably with `PersistentView`s as well.  I'd agree that the index doesn't support this query.  

If you add that index by hand in your testing environment, does it alleviate all the problems?  If not, what remains?

**Comment by michaelom** [comment]
> Which driver are you using?

Casbah 2.8.1
> Has the journal been upgraded (journal auto migrate) already?

Yes.

This is new new "Schema":
```
{
  "_id": ObjectId("5680f3f011704407e4547f3f"),
  "pid": "Account",
  "from": NumberLong(95273),
  "to": NumberLong(95273),
  "events": [
    {
      "v": NumberInt(1),
      "pid": "Account",
      "sn": NumberLong(95273),
      "_t": "repr",
      "p": "<Mongo Binary Data>",
      "_w": "80aef660-ac35-479d-8666-1cfbd686eed8"
    }
  ],
  "v": NumberInt(1)
}
```

> ...I believe it runs as part of the recovery phase for PersistentActors, and presumably with PersistentViews as well....

Yes, PersistentActors and PersistentViews should behave the same with autoRecovery on.
That means that I should see this Query exactly once during actor start. In fact I see it every second.


Meanwhile we added an index to `to`. (like you suggested). We added the index by hand and now everything seems to be the way it should.

We also like to believe we know why it works now: :-) 
When no descending index is provided and the query runs against a big collection, then the query results in an error: 
```
db.events.find({ $query: { pid: "Account" }, $orderby: { to: -1 } })
Error: error: {
    "$err" : "Executor error: Overflow sort stage buffered data usage of 33554553 bytes exceeds internal limit of 33554432 bytes",
    "code" : 17144
}
```
The documentation is very clear:
> If MongoDB cannot use an index to get documents in the requested sort order, the combined size of all documents in the sort operation, plus a small overhead, must be less than 32 megabytes.

So I assume the PersistentView was attempting an initial recovery and failed due to a missing index (or the collection size I you see it that way). The failed attempt caused another attempt which failed again and so on...

**Comment by michaelom** [comment]
I guess for most PersistentActors/-View this will not be an issue as long as the event stream for a given `pid` is small enough. Still I think the index should be created by your plugin. We used the following index:

```[
    {
        "v" : 1,
        "key" : {
            "_id" : 1
        },
        "name" : "_id_",
        "ns" : "account.events"
    },
    {
        "v" : 1,
        "unique" : true,
        "key" : {
            "pid" : 1,
            "from" : 1,
            "to" : 1
        },
        "name" : "events_index",
        "ns" : "account.events"
    },
    {
        "v" : 1,
        "key" : {
            "to" : -1
        },
        "name" : "to_-1",
        "ns" : "account.events"
    }
]```

Do you agree?

**Comment by scullxbones** [comment]
Yep that sounds plausible, persistence error happens with every `Update` message that is auto scheduled @ 1s intervals.

Good news on the index, i'll add it - I agree it should be in the plugin.

**Comment by michaelom** [comment]
Thanks again for your _near time_ support. I let you know how everything turned out.

**Comment by scullxbones** [comment]
Fixed by `1.1.4`.  Thanks for helping me troubleshoot!
